site3/website/src/pages/bps/BP-62-new-API-for-batched-reads.md

```diff
@@ -0,0 +1,305 @@
+# BP-62 New API for batched reads
+
+# Motivation
+
+The BookKeeper client has provided an API [0] that allows for the efficient reading of entries [1] from bookies. 
+While users can specify a start and end entry ID for a ledger [2] according to the API's definition, 
+the reality is that the bookie server can only read a single entry at a time. Consequently, the BookKeeper client is 
+required to send a request for each entry to the bookies based on the user-specified start and end entry IDs. 
+This approach is highly inefficient and can result in significant CPU resource consumption, especially when reading a 
+large number of entries. The frequent RPC calls to the bookies also put a heavy burden on the bookie server, which must 
+process read requests from the IO thread and subsequently jump to the entry read thread pool.
+![single_read.png](../../../static/img/single_read.png)
+
+
+A more effective way to read entries is to use one RPC request to read multiple entries at once. This approach will 
+significantly reduce the number of RPC calls, thread jumps, and enqueue and dequeue operations of the thread pool.
+![batch_read.png](../../../static/img/batch_read.png)
+
+To demonstrate this point, I have conducted a simple benchmark test using the BookKeeper perf tool to compare batch 
+and individual entry readings. The test case was repeated to read the data from the same ledger and to avoid the impact 
+of disk performance on test results, the entries were read from the cache since the batch read protocol change would not 
+introduce any disk read improvements. The results show that using the batched entry read API provides a remarkable 
+10x performance improvement.
+
+Here is the output of the BookKeeper perf tool with ensemble=1, write=1, and ack=1.
+```text
+Batch(100): Read 1000100 entries in 8904ms
+Batch(500): Read 1000500 entries in 12182ms
+Non-Batch: Read 1000130 entries in 199928ms
+```
+The key difference with the batch read is that the BookKeeper client sends only 10,000 requests to the server 
+(compared to 1,000,000 in non-batch read), and the server sends 10,000 responses back to the client (versus 1,000,000 in non-batch read).
+This approach reduces the number of RPC calls made and eliminates the need for the server to process a large number of 
+requests from the IO thread and the entry read thread pool, resulting in a significant improvement in performance.
+
+Hence, the objective of this proposal is to enhance the performance of entry reading by introducing a batch entry reading protocol 
+that takes into account the expected count and size of entries.
+- **Optimize entry reading performance:** By reading multiple entries in a single RPC request, the network communication and RPC call overhead can be reduced, thereby optimizing the reading performance.
```

Comment:
We should also add the effect of 1 failed read operation in a batch of 100 (eg: because the bookie rejects the operation when the executor queue is full), where the entire batch is failed in the BK client and we have to perform the entire batch read again.

Reply:
See section `Loose constraints`. If the batch read request read entries [0-100]. Then the bookie received the request, If reading entries [0-50] is successful, but reading entry 51 fails, we will return entries [0-50] to the client and trigger the callback. The data returned to the user will only include entries [0-50]. The user will need to read the data starting from entry 51 in the next attempt.